cirq/work/observable_measurement.py

```diff
@@ -211,3 +269,148 @@ def _subdivide_meas_specs(
 
     return flippy_mspecs, repetitions
+
+
+def _to_sweep(param_tuples):
+    """Turn param tuples into a sweep."""
+    to_sweep = [dict(pt) for pt in param_tuples]
+    to_sweep = study.to_sweep(to_sweep)
+    return to_sweep
+
+
+def _get_qubits(max_settings: Iterable[InitObsSetting]):
+    """Helper function to find all the qubits in a suite of settings used
+    in `measure_grouped_settings`.
+
+    Qubits are returned in sorted order.
+    """
+    qubits = set()
+    for max_setting in max_settings:
+        max_in_st = max_setting.init_state
+        qubits |= set(max_in_st.qubits)
+    return sorted(qubits)
+
+
+def _determine_needs_init_layer(init_state: value.ProductState):
+    """Helper function to determine whether we need the initial layer of
+    rotations used in `measure_grouped_settings`.
+    """
+    for _, st in init_state:
+        if st != value.KET_ZERO:
+            return True
+
+    return False
+
+
+def measure_grouped_settings(
+    circuit: 'cirq.Circuit',
+    grouped_settings: Dict[InitObsSetting, List[InitObsSetting]],
+    sampler: 'cirq.Sampler',
+    desired_repetitions: int,
+    *,
+    readout_symmetrization: bool = False,
+    circuit_sweep: 'cirq.study.sweepable.SweepLike' = None,
+) -> List[BitstringAccumulator]:
+    """Measure a suite of grouped InitObsSetting settings.
+
+    This is a low-level API for accessing the observable measurement
+    framework. See also `measure_observables` and `measure_observables_df`.
+
+    Args:
+        circuit: The circuit. This can contain parameters, in which case
+            you should also specify `circuit_sweep`.
+        grouped_settings: A series of setting groups expressed as a dictionary.
+            The key is the max-weight setting used for preparing single-qubit
+            basis-change rotations. The value is a list of settings
+            compatible with the maximal setting you desire to measure.
+            Automated routing algorithms like `group_settings_greedy` can
+            be used to construct this input.
+        sampler: A sampler.
+        desired_repetitions: How many repetitions per observable.
+        readout_symmetrization: If set to True, each `meas_spec` will be
+            split into two runs: one normal and one where a bit flip is
+            incorporated prior to measurement. In the latter case, the
+            measured bit will be flipped back classically and accumulated
+            together. This causes readout error to appear symmetric,
+            p(0|0) = p(1|1).
+        circuit_sweep: Additional parameter sweeps for parameters contained
+            in `circuit`. The total sweep is the product of the circuit sweep
+            with parameter settings for the single-qubit basis-change rotations.
+    """
+
+    qubits = _get_qubits(grouped_settings.keys())
+    qubit_to_index = {q: i for i, q in enumerate(qubits)}
+
+    needs_init_layer = False
+    for max_setting in grouped_settings.keys():
```

Comment:
we could have the whole (for settings as well) "needs init layer" logic extracted into a private method, not just the state part of it

Reply:
what do you mean? in any event, I inlined the now-one-line definition of `_[determine_]needs_init_layer`.

Whether an init layer is required only depends on the init state.

Reply:
I meant having `_determine_needs_init_layer` separately seemed like it could have encompassed these (now) 5 lines of code. This method is very long, I was trying to chip away extractable pieces so it reads better.

Reply:
factored out into a helper function, thanks